x-pack/plugins/apm/server/routes/settings/anomaly_detection.ts

```diff
@@ -5,27 +5,62 @@
  */
 
 import * as t from 'io-ts';
+import { ErrorCode } from '../../../common/anomaly_detection';
+import { PromiseReturnType } from '../../../typings/common';
+import { InsufficientMLCapabilities } from '../../../../ml/server';
 import { createRoute } from '../create_route';
 import { getAnomalyDetectionJobs } from '../../lib/anomaly_detection/get_anomaly_detection_jobs';
 import { createAnomalyDetectionJobs } from '../../lib/anomaly_detection/create_anomaly_detection_jobs';
 import { setupRequest } from '../../lib/helpers/setup_request';
 import { getAllEnvironments } from '../../lib/environments/get_all_environments';
 import { hasLegacyJobs } from '../../lib/anomaly_detection/has_legacy_jobs';
+import { AnomalyDetectionError } from '../../lib/anomaly_detection/anomaly_detection_error';
+
+type Jobs = PromiseReturnType<typeof getAnomalyDetectionJobs>;
 
 // get ML anomaly detection jobs for each environment
 export const anomalyDetectionJobsRoute = createRoute(() => ({
   method: 'GET',
   path: '/api/apm/settings/anomaly-detection',
   handler: async ({ context, request }) => {
     const setup = await setupRequest(context, request);
-    const [jobs, legacyJobs] = await Promise.all([
-      getAnomalyDetectionJobs(setup, context.logger),
-      hasLegacyJobs(setup),
-    ]);
-    return {
-      jobs,
-      hasLegacyJobs: legacyJobs,
-    };
+
+    try {
+      const [jobs, legacyJobs] = await Promise.all([
+        getAnomalyDetectionJobs(setup, context.logger),
+        hasLegacyJobs(setup),
+      ]);
+      return {
+        jobs,
+        hasLegacyJobs: legacyJobs,
+      };
+    } catch (e) {
+      // ML error
+      if (e instanceof InsufficientMLCapabilities) {
+        return {
+          jobs: [] as Jobs,
+          hasLegacyJobs: false,
+          errorCode: ErrorCode.MISSING_READ_PRIVILEGES,
+        };
+      }
+
+      // AnomalyDetectionError error
+      if (e instanceof AnomalyDetectionError) {
+        return {
+          jobs: [] as Jobs,
+          hasLegacyJobs: false,
+          errorCode: e.code,
+        };
+      }
+
+      // unexpected error
+      context.logger.warn(e.message);
+      return {
+        jobs: [] as Jobs,
+        hasLegacyJobs: false,
+        errorCode: ErrorCode.UNEXPECTED,
+      };
+    }
```

Comment:
You could create a new function which returns the ML error code, then you could also use it on `createAnomalyDetectionJobsRoute`.
```suggestion
      const mlErrorCode = getMLErrorCode(e);
      return {
          jobs: [] as Jobs,
          hasLegacyJobs: false,
          errorCode: mlErrorCode
      }
```

Reply:
Good idea 👍